Checkout: Make sure the font size for PMME is smaller than the labels

danielmx-dev opened the pull request:
Fixes #9818

#### Changes proposed in this Pull Request

The size of PMME inside the Payment labels should be smaller than the label that contains the payment method name.

#### Testing instructions

* You can use the Storefront theme
* Activate Klarna, Affirm, and/or Afterpay.
* Add a product over 50 USD and go to blocks checkout.
* To compare between before and after, you'll need to clear the transients. You can use the WCPay devtools or use this query:
```sql
DELETE FROM `wp_options` WHERE `option_name` LIKE '_transient_wcpay_upe_%' OR `option_name` LIKE '_transient_wcpay_wc_blocks_upe_appearance%';
```

| Before | After |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/e4f1d5ce-d989-4dfd-b2d8-cf4a5c7c0ee9) | ![image](https://github.com/user-attachments/assets/96dcaf67-4405-4e9f-8cc9-0536efd16196) | 

* Go to classic/shortcode checkout and also compare the results

| Before | After |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/9ac0c851-fd82-4681-90b4-e14374d9c1c9) | ![image](https://github.com/user-attachments/assets/993132f3-2e7c-4929-984b-6a9b1ad64d46) | 

**Regression Testing**
No changes in size should be visible for PMME in Product Details and Cart pages.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
Checkout: Make sure the font size for PMME is smaller than the labels

Commit:
Add changelog

Commit:
Preserve font size in floating labels

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9925` or branch name `fix/change-font-size-pmme-checkout` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/change-font-size-pmme-checkout) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 0025fcbd0ec6b477b1a56308e5e2d6caa15676ad
- Build time: 2025-01-06 14:38:11 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Commit:
Update tests and skip the resizing if the fontSize is missing

danielmx-dev commented:
@pierorocca Can you please confirm that the adjustment in size looks fine in the screenshots provided? I set the PMME font to be at most 85% of the Payment Method Label size in both blocks and shortcode checkout. Other flows (product details, cart) are unchanged.

danielmx-dev committed:
Merge branch 'develop' into fix/change-font-size-pmme-checkout

FangedParakeet commented on client/checkout/upe-styles/index.js:
```diff
@@ -527,9 +565,18 @@ export const getAppearance = ( elementsLocation, forWooPay = false ) => {
+		globalRules.fontSizeBase = ensureFontSizeSmallerThan(
+			selectors.pmmeRelativeTextSizeSelector,
+			paragraphRules.fontSize,
+			PMME_RELATIVE_TEXT_SIZE // The font size of the payment method messaging should be 85% or less of the target font size.
```
Just a thought, but it's a little strange to pass a constant as a function parameter. This function isn't used anywhere else and there isn't yet really an indication that it might be, so this parameter seems a little superfluous, since it will never change.

Perhaps we can remove the parameter and just use this constant inside the body of `ensureFontSizeSmallerThan()` instead? 🤔

FangedParakeet commented on client/checkout/upe-styles/index.js:
```diff
@@ -527,9 +565,18 @@ export const getAppearance = ( elementsLocation, forWooPay = false ) => {
 		colorBackground: backgroundColor,
 		colorText: paragraphRules.color,
 		fontFamily: paragraphRules.fontFamily,
+		// fontSizeBase: transformFontSize( paragraphRules.fontSize, 0.9 ),
```
```suggestion
```
I presume we don't need this anymore.

FangedParakeet reviewed (changes requested):
I've left a few small comments for tiny things that could be tweaked. In general, I thought the changes here look pretty tidy: I think the appearance looks perfect on the blocks checkout, though I think the font-size of the PMME text on the shortcode checkout appeared a little small.

Here's what the [Figma designs](https://www.figma.com/design/g1bWnb6JBO3X3bGrgJnRJe/Agnostic-Payment-Component?node-id=154-7168&node-type=frame&t=QY4NJiFRTM3BPR86-0) look like. 

### Figma
#### Blocks
![Figma blocks design](https://cdn-std.droplr.net/files/acc_1104206/Pd0L2N)
_Label: 15px, PMME: 13px, 87% difference_

#### Shortcode
![Figma shortcode design](https://cdn-std.droplr.net/files/acc_1104206/qkRO8t)
_Label: 16px, PMME: 13px, 81% difference_

Following the sizes described in the designs, I can see how you arrived at using an 85% multiplier for the font shrinkage. However, here's what I noticed testing across a few themes locally.

### Storefront
#### Blocks
![Storefront blocks PMME](https://cdn-std.droplr.net/files/acc_1104206/hYO3xC)
_Looks fine_

#### Shortcode
![Storefront shortcode PMME](https://cdn-std.droplr.net/files/acc_1104206/ztxWZ8)
_Looks a little small_

AFAICT, for both blocks and shortcode the target font-size is 16px and the computed PMME font-size becomes 13.6px, but for some reason this appears slightly differently on the shortcode checkout.

Now here come a few of my favourite uglier themes for further comparison. 

### Twenty Twenty-One
#### Blocks

![2021 Blocks PMME](https://cdn-std.droplr.net/files/acc_1104206/pgSeoz)
_Looks OK_

Here the label text is supposedly 20px and our PMME text becomes 17px.

#### Shortcode
![2021 shortcode PMME](https://cdn-std.droplr.net/files/acc_1104206/hiXpLi)
_This looks a little too small_

The label font here is 16px and our PMME font is 13.6px.

### Deli
#### Blocks
![Deli blocks PMME](https://cdn-std.droplr.net/files/acc_1104206/eD0buF)
_Looks fine_

13.712 px ➡️ 11.6552px

#### Shortcode 
![Deli shortcode PMME](https://cdn-std.droplr.net/files/acc_1104206/SumBOe)
_This looks OK too actually, though maybe touching on the smaller side_

13.712 px ➡️ 11.6552px

The more that I stare at this, I think that the main issue might be the difference in spacing on the shortcode checkout. 

One other thing to note that in the designs, while the shortcode text has a greater font-size differential, the shortcode design actually uses a different font from the blocks design. Perhaps this explains the difference in appearance. 

Anyways, let me know your thoughts and whether you noticed any of these sizing differences yourself or am I just seeing things? Do we maybe need to use a different `PMME_RELATIVE_TEXT_SIZE` constant for the blocks and shortcode checkouts? Let me know what you think. 🙏 

CC @pierorocca

Commit:
Update PMME_RELATIVE_TEXT_SIZE and add more selectors to target secondary font sizes

Commit:
Pass fontRules to PMME container in blocks checkout

danielmx-dev commented:
@FangedParakeet I have updated the default size percentage to `.875` so it matches the 16px/14px ratio. I also added more selectors in an attempt to match elements using a secondary font size during Checkout as Piero mentioned.

Another thing that I noticed is that even if now both Classic and Blocks checkout were using 14px in the PMME containers, the text looked a bit larger in blocks. I found out that we were not sending the font rules to the container, so it used a font (Helvetica Neue) that looked larger than the actual font (Source Sans Pro).

After fixing that, now the fields look like this (Storefront):

| Blocks | Shortcode |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/749d2a6e-1b41-440a-a144-ffb8f64fa4ac) | ![image](https://github.com/user-attachments/assets/9a0c54e9-d56f-411e-8f6d-ae874838956a) |

pierorocca commented:
One of the secondary goals here is to have the fonts across the entire checkout form aligned. So if the payments secondary font size ends up being computed differently to the secondary font size in order summary component or elsewhere on Checkout that would be a problem and an inconsistency.

danielmx-dev commented:
> One of the secondary goals here is to have the fonts across the entire checkout form aligned. So if the payments secondary font size ends up being computed differently to the secondary font size in order summary component or elsewhere on Checkout that would be a problem and an inconsistency.

I added new selectors to target the mentioned secondary font sizes (if available). 

For classic checkout I wasn't able to find many alternative sizes, so I chose the `#payment .payment_methods li .payment_box fieldset` selector, which applies the font-size for this element:
![image](https://github.com/user-attachments/assets/f6f04a50-d9dc-46e3-a8f3-e9ae63a83fb8)

For blocks checkout I went with `.wc-block-components-checkout-step__description`, which should target this element:
![image](https://github.com/user-attachments/assets/86ace2fa-69a7-491f-ab45-88efd4967c64)

If those elements are not available it will fallback to other elements in the page that contain text. If the resulting size is too large, then it will reduce the size using the 0.875 ratio.

Commit:
Handle NaN font sizes

pierorocca commented:
@elizaan36 @ralucaStan  could you please sanity check this against pdFofs-2xh-p2? Thank you.

elizaan36 commented:
@danielmx-dev Could you share before and after images after [this change](https://github.com/Automattic/woocommerce-payments/pull/9925#issuecomment-2551935301). Thank you.

danielmx-dev commented:
> @danielmx-dev Could you share before and after images after [this change](https://github.com/Automattic/woocommerce-payments/pull/9925#issuecomment-2551935301). Thank you.

The Before remains unchanged as the one shown in the description.

Blocks
| Before | After |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/e4f1d5ce-d989-4dfd-b2d8-cf4a5c7c0ee9) | ![image](https://github.com/user-attachments/assets/749d2a6e-1b41-440a-a144-ffb8f64fa4ac) | 


Shortcode
| Before | After |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/9ac0c851-fd82-4681-90b4-e14374d9c1c9) | ![image](https://github.com/user-attachments/assets/9a0c54e9-d56f-411e-8f6d-ae874838956a) |

danielmx-dev committed:
Merge branch 'develop' into fix/change-font-size-pmme-checkout

pierorocca commented:
Are the save payment methods in this location new? I don't recall seeing them located within the card component before.
![image](https://github.com/user-attachments/assets/992c8fda-0fc2-4317-9f11-87989a42d98a)

pierorocca commented:
Existing:

![image](https://github.com/user-attachments/assets/79752983-119e-41a3-a71e-8b7a096a142b)

danielmx-dev commented:
> Are the save payment methods in this location new? I don't recall seeing them located within the card component before. ![image](https://private-user-images.githubusercontent.com/109928953/398243412-992c8fda-0fc2-4317-9f11-87989a42d98a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUwNDg2NTMsIm5iZiI6MTczNTA0ODM1MywicGF0aCI6Ii8xMDk5Mjg5NTMvMzk4MjQzNDEyLTk5MmM4ZmRhLTBmYzItNDMxNy05ZjExLTg3OTg5YTQyZDk4YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyNFQxMzUyMzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YWMxMGU3NGU0MDJlZjM5MjI5MzQyNGZhNjA5ZWQyZDM0NWM4NjQ3MzkyNTUwZWNmODQ1ZTk0NmRmYjFmN2JlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.iVnPXKqIeLHpyCsqEDzU2zK2ohYQ79o7x2UqzQ_9BxM)

This is for Classic/Shortcode Checkout, and it is existing behavior.

pierorocca commented:
I see that now thanks Daniel.

elizaan36 commented:
Thanks @danielmx-dev for sharing the before and after images. The text size changes look good to me!

ralucaStan commented:
If there is a change needed for the checkout shortcode please open a separate ticket on https://github.com/woocommerce/woocommerce/issues

danielmx-dev committed:
Merge branch 'develop' into fix/change-font-size-pmme-checkout

FangedParakeet commented on client/checkout/blocks/payment-method-label.js:
```diff
@@ -57,6 +57,8 @@ export default ( { api, title, countries, iconLight, iconDark, upeName } ) => {
 		getUPEConfig( 'wcBlocksUPEAppearanceTheme' )
 	);
 
+	const fontRules = useMemo( () => getFontRulesFromPage(), [] );
```
```suggestion
	const fontRules = useMemo( getFontRulesFromPage, [] );
```
Not sure if this is even an enhancement or if this is unconventional and might have longer term drawbacks, if we decide to add some parameters to `getFontRulesFromPage`, but since `getFontRulesFromPage` currently requires no function parameters, this could save us a pocket-full of characters. 🤷 

Feel free to ignore, as perhaps leaving this as it is currently is more conventional for usage with `useMemo`.

danielmx-dev replied:
Thanks for the suggestion but I think I'll just keep the current implementation as it is more explicit and conventional, as you mentioned.

FangedParakeet reviewed (approved):
Cheers for the tweaks and updates here! 🙌 

Checked the changes across a few of my favourite and least favourite themes and the PMME text size appears as desired in the designs. Left one tiny comment, but it's pretty negligible, so feel free to move on without it. 🚢
